pylearner.ts

```diff
@@ -1990,7 +2188,92 @@ function parseProofOutline(stmts: Statement[], i: number, precededByAssert: bool
+  let mayAliasRelationBlock = performMayAliasAnalysis(blockStatement.stmts, 0, previousStatementMaybeAliasRelation)!;
+  const scopeBlock = blockStatement.scope!;
+  const localVars = ImmutableSet(Object.keys(scopeBlock!.bindings));
+  const outerscopeVars = getOuterScopeVariables(scopeBlock.outerScope!);
+  let varsToKeep = outerscopeVars.subtract(localVars);
+  const mayAliasRelationResult = mayAliasRelationBlock.keepOnlyVariables(varsToKeep);
+  return mayAliasRelationResult;
+} 
+
+function performMayAliasAnalysis(stmts: Statement[], i: number, previousStatementMayAliasRelation: MayAliasRelation): MayAliasRelation {
+  if (!(stmts.length == i)) {
+    let stmt = stmts[i];
+    stmt.mayAliasRelation = previousStatementMayAliasRelation;
+    if (stmt instanceof ExpressionStatement) {
+      if (stmt.expr instanceof AssignmentExpression && stmt.expr.op == '=' && stmt.expr.lhs instanceof VariableExpression) {
+        if (stmt.expr.rhs instanceof VariableExpression)
+          stmt.mayAliasRelation = stmt.mayAliasRelation.addMayAliasLink(stmt.expr.lhs.name, stmt.expr.rhs.name);
+        else
+          stmt.mayAliasRelation = stmt.mayAliasRelation.removeVariable(stmt.expr.lhs.name);
+      }
+    } else if (stmt instanceof IfStatement) {
+      if (stmt.thenBody instanceof BlockStatement && stmt.elseBody instanceof BlockStatement &&  stmt.elseBody != null && stmt.thenBody != null) {
+        let mayAliasRelationThenBlock = performMayAliasAnalysisOnBlock(stmt.thenBody, stmt.mayAliasRelation);
+        let mayAliasRelationElseBlock = performMayAliasAnalysisOnBlock(stmt.elseBody, stmt.mayAliasRelation);
+        let combinedMayAliasRelation = mayAliasRelationThenBlock.combineWithMayAliasRelation(mayAliasRelationElseBlock);
+        return performMayAliasAnalysis(stmts, i+1, combinedMayAliasRelation);
+      }
+    } else if (stmt instanceof WhileStatement) {
+      if (stmt.body instanceof BlockStatement && stmt.body != null) {
+        const whileBlock = stmt.body;
+        let mayAliasRelationWhileBlock = performMayAliasAnalysisOnBlock(whileBlock, stmt.mayAliasRelation);
+        let mayAliasRelationWhileBlockCheck = performMayAliasAnalysisOnBlock(whileBlock, mayAliasRelationWhileBlock);
+        while (!mayAliasRelationWhileBlock.equals(mayAliasRelationWhileBlockCheck)) {
+          mayAliasRelationWhileBlock = mayAliasRelationWhileBlockCheck;
+          mayAliasRelationWhileBlockCheck = performMayAliasAnalysisOnBlock(whileBlock, mayAliasRelationWhileBlock);
+        }
+        return performMayAliasAnalysis(stmts, i+1, mayAliasRelationWhileBlockCheck);
+      }
+    }
+    return performMayAliasAnalysis(stmts, i+1, stmt.mayAliasRelation);
+  }
+  else
+    return previousStatementMayAliasRelation;
+}
+
 function checkProofOutline(checkEntailments: boolean, total: boolean, env: Env_, stmts: Statement[]) {
+  performMayAliasAnalysis(stmts, 0, new MayAliasRelation());
```

Comment:
We gaan er hier van uit dat er geen aliasing is tussen de invoervariabelen van het bewezen stuk code; dit is echter niet per se het geval. Om helemaal veilig te zijn moeten we hier alle variabelen van type `list` die hier in scope zijn beschouwen als mogelijks aliases van elkaar, denk ik. (Het kan trouwens geen kwaad om gewoon alle variabelen in scope, ongeacht hun type, te beschouwen als mogelijke aliases, denk ik.)

Reply:
Lijkt inderdaad veiliger, alleen ben ik nog beetje op zoek hoe ik aan die scope geraak. In de methode checkProofOutline zelf geraak je er niet aan denk ik. In de methode checkProofOutlines die de checkProofOutline oproept kan je wel aan de parameters. Die zou je dan kunnen meegeven in checkProofOutline en initialiseren in de may-alias-relation. Maar hoe ik aan die hele scope geraak, ben ik nog mee bezig.

Reply:
Je kan bv. in de `check`-methode van klasse AssertStatement de meegegeven scope opslaan in een attribuut van het AssertStatement-object.